olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/TileRequest.h

```diff
@@ -0,0 +1,212 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+#pragma once
+
+#include <boost/optional.hpp>
+#include <sstream>
+#include <string>
+
+#include <olp/core/geo/tiling/TileKey.h>
+#include <olp/dataservice/read/FetchOptions.h>
+
+namespace olp {
+namespace dataservice {
+namespace read {
+
+/**
+ * @brief Encapsulates the fields required to request tile for the given
+ * key.
+ *
+ * You should specify tile key. Additionaly offset and depth could be set. If
+ * not set used default values, for offset default value is 2, for depth - 4.
+ */
+class DATASERVICE_READ_API TileRequest final {
+ public:
+  /**
+   * @brief Gets the billing tag to group billing records together.
+   *
+   * The billing tag is an optional free-form tag that is used for grouping
+   * billing records together. If supplied, it must be 4–16 characters
+   * long and contain only alphanumeric ASCII characters [A-Za-z0-9].
+   *
+   * @return The `BillingTag` string or `boost::none` if the billing tag is not
+   * set.
+   */
+  inline const boost::optional<std::string>& GetBillingTag() const {
+    return billing_tag_;
+  }
+
+  /**
+   * @brief Sets the billing tag for the request.
+   *
+   * @see `GetBillingTag()` for information on usage and format.
+   *
+   * @param tag The `BillingTag` string or `boost::none`.
+   *
+   * @return A reference to the updated `TileRequest` instance.
+   */
+  inline TileRequest& WithBillingTag(const boost::optional<std::string>& tag) {
+    billing_tag_ = tag;
+    return *this;
+  }
+
+  /**
+   * @brief Sets the billing tag for the request.
+   *
+   * @see `GetBillingTag()` for information on usage and format.
+   *
+   * @param tag The rvalue reference to the `BillingTag` string or
+   * `boost::none`.
+   *
+   * @return A reference to the updated `TileRequest` instance.
+   */
+  inline TileRequest& WithBillingTag(std::string&& tag) {
+    billing_tag_ = std::move(tag);
+    return *this;
+  }
+
+  /**
+   * @brief Sets the tile key for the request.
+   *
+   * @param tile_key Tile key value.
+   *
+   * @return A reference to the updated `TileRequest` instance.
+   */
+  inline TileRequest& WithTileKey(geo::TileKey&& tile_key) {
+    tile_key_ = std::move(tile_key);
+    return *this;
+  }
+
+  /**
+   * @brief Sets the tile key for the request.
+   *
+   * @param tile_key Tile key value.
+   *
+   * @return A reference to the updated `TileRequest` instance.
+   */
+  inline TileRequest& WithTileKey(const geo::TileKey& tile_key) {
+    tile_key_ = tile_key;
+    return *this;
+  }
```

Comment:
I think only one with pass by value should be enough, considering that TileKey is nothing more then a coupla of ints.

```cpp
inline TileRequest& WithTileKey(geo::TileKey tile_key)
```